druid-shell/src/alert.rs

```diff
@@ -0,0 +1,451 @@
+/// If there is some piece of information that the user must know immediately,
+/// e.g. that storage space is running very low and will soon cause a critical error,
+/// or if the storage space has already depleted and the critcial error has happened.
+/// Alerts are also useful when there is some decision that the user must make
+/// and the application can't continue without making that decision. Even so
+/// you should think hard when designing your application whether you can have
+/// reasonable default behavior which can happen without alerts. It's better to
+/// offer an *undo* feature than to ask for confirmation on deletion.
+///
+/// The fewer alerts the user encounters, the better their experience.
+///
+/// # Modality
+///
+/// Alerts can operate in two different modality modes. By default
+/// they are window scoped. Alternatively they can be application scoped.
+///
+/// While an alert dialog is open it takes interaction priority in its modality scope
+/// and the user won't be able to continue interacting with anything else.
+/// In window scope this means the user can't interact with the parent window
+/// and in application scope this means the user can't interact with any other window.
+///
+/// # Multiple alerts
+///
+/// There is no guaranteed behavior when multiple alert dialogs are requested with
+/// overlapping modality scope. Depending on platform specifics, all the alerts
+/// might be visible or just some of them. Regardless of visibility, only one of
+/// the alert dialogs will be interactable. When that alert dialog is closed,
+/// another one will become interactable.
+///
+/// If the modality scopes don't overlap, i.e. when multiple alerts are originating
+/// from different windows and they are all window scoped, all the alerts can be
+/// interacted with and they won't block eachother.
+#[derive(Debug, Clone)]
+pub struct AlertOptions {
+    /// Whether the alert is app-modal.
+    pub(crate) app_modal: bool,
+    /// The context of the alert.
+    pub(crate) context: String,
+    /// The primary message of the alert.
+    pub(crate) message: String,
+    /// The supplemental text of the alert.
+    pub(crate) description: String,
+    /// The icon of the alert.
+    pub(crate) icon: Option<AlertIcon>,
+    /// The primary button of the alert.
+    pub(crate) primary: AlertButton,
+    /// The cancel button of the alert.
+    pub(crate) cancel: Option<AlertButton>,
+    /// The alternative buttons of the alert.
+    pub(crate) alternatives: Vec<AlertButton>,
+}
+
+/// Alert dialog icon.
+///
+/// # macOS
+///
+/// On macOS the information and warning icons are the same.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum AlertIcon {
+    /// The alert dialog is providing important time sensitive information.
+    Information,
+    /// The alert dialog is warning about a potential future problem.
+    Warning,
+    /// The alert dialog is informing of a critical problem that has already happened.
+    Error,
```

Comment:
I would mark this as non-exhaustive. In particular we will likely want at some point the option of supplying a custom icon?

Reply:
Yeah custom icons could come later. However I'm not sure why anyone would be matching on this enum, as we only take it as input and never output it. So right now it would just create an additional useless match arm for `druid-shell` itself.

Reply:
I guess this is more a philosophical point; if this is exposed publicly then any addition would be a semvar major change. Maybe not a practical concern right now, but something I like to keep in mind.